File: src/selfrionette/schemas/viewer_control_message.py
from __future__ import annotations

import json
from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Literal

ViewerControlSourceKind = Literal["keyboard", "gamepad"]
ViewerControlEnvelopeType = Literal["viewer_control_message"]
ViewerControlKeyboardFocusState = Literal["focused", "blurred"]


class ViewerControlMessageError(ValueError):
    pass


@dataclass(frozen=True, slots=True)
class ViewerControlKeyboardMessage:
    active_key_codes: tuple[str, ...] = ()
    key_state: Mapping[str, bool] = field(default_factory=dict)
    focus_state: ViewerControlKeyboardFocusState | None = None
    zero_state: bool | None = None


@dataclass(frozen=True, slots=True)
class ViewerControlGamepadButtonMessage:
    pressed: bool
    value: float | None = None


@dataclass(frozen=True, slots=True)
class ViewerControlGamepadMessage:
    connected: bool
    index: int | None = None
    id: str | None = None
    axes: tuple[float, ...] = ()
    buttons: tuple[ViewerControlGamepadButtonMessage, ...] = ()
    stale: bool | None = None
    zero_state: bool | None = None


@dataclass(frozen=True, slots=True)
class ViewerControlMessage:
    type: ViewerControlEnvelopeType
    timestamp_s: float
    source_kind: ViewerControlSourceKind
    sequence: int | None = None
    keyboard: ViewerControlKeyboardMessage | None = None
    gamepad: ViewerControlGamepadMessage | None = None
    metadata: Mapping[str, object] = field(default_factory=dict)


def _is_plain_mapping(value: object) -> bool:
    return isinstance(value, Mapping)


def _is_str(value: object) -> bool:
    return isinstance(value, str)


def _is_bool(value: object) -> bool:
    return isinstance(value, bool)


def _is_int(value: object) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def _is_finite_number(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool) and value == value and value not in {float("inf"), float("-inf")}


def _ensure_allowed_keys(payload: Mapping[str, object], allowed_keys: set[str], context: str) -> None:
    unknown_keys = sorted(set(payload) - allowed_keys)
    if unknown_keys:
        unknown = ", ".join(unknown_keys)
        raise ViewerControlMessageError(f"{context} contains unknown fields: {unknown}")


def _coerce_string_tuple(value: object, *, context: str) -> tuple[str, ...]:
    if not isinstance(value, list) or any(not _is_str(item) for item in value):
        raise ViewerControlMessageError(f"{context} must be an array of strings")
    return tuple(value)


def _coerce_number_tuple(value: object, *, context: str) -> tuple[float, ...]:
    if not isinstance(value, list) or any(not _is_finite_number(item) for item in value):
        raise ViewerControlMessageError(f"{context} must be an array of finite numbers")
    return tuple(float(item) for item in value)


def _coerce_key_state(value: object, *, context: str) -> dict[str, bool]:
    if not _is_plain_mapping(value):
        raise ViewerControlMessageError(f"{context} must be a JSON object")

    key_state: dict[str, bool] = {}
    for key, mapped_value in value.items():
        if not _is_str(key):
            raise ViewerControlMessageError(f"{context} keys must be strings")
        if not _is_bool(mapped_value):
            raise ViewerControlMessageError(f"{context}[{key!r}] must be a boolean")
        key_state[key] = mapped_value

    return key_state


def _coerce_optional_bool(value: object, *, context: str) -> bool | None:
    if value is None:
        raise ViewerControlMessageError(f"{context} must be a boolean")
    if not _is_bool(value):
        raise ViewerControlMessageError(f"{context} must be a boolean")
    return value


def _coerce_optional_focus_state(value: object, *, context: str) -> ViewerControlKeyboardFocusState | None:
    if value is None:
        raise ViewerControlMessageError(f"{context} must be 'focused' or 'blurred'")
    if value not in {"focused", "blurred"}:
        raise ViewerControlMessageError(f"{context} must be 'focused' or 'blurred'")
    return value


def _coerce_optional_number(value: object, *, context: str) -> float | None:
    if value is None:
        raise ViewerControlMessageError(f"{context} must be a finite number")
    if not _is_finite_number(value):
        raise ViewerControlMessageError(f"{context} must be a finite number")
    return float(value)


def _coerce_gamepad_button_message(value: object, *, context: str) -> ViewerControlGamepadButtonMessage:
    if not _is_plain_mapping(value):
        raise ViewerControlMessageError(f"{context} must be a JSON object")

    allowed_keys = {"pressed", "value"}
    _ensure_allowed_keys(value, allowed_keys, context)

    if "pressed" not in value:
        raise ViewerControlMessageError(f"{context}.pressed is required")
    if not _is_bool(value["pressed"]):
        raise ViewerControlMessageError(f"{context}.pressed must be a boolean")

    pressed = value["pressed"]
    button_value = _coerce_optional_number(value["value"], context=f"{context}.value") if "value" in value else None

    return ViewerControlGamepadButtonMessage(pressed=pressed, value=button_value)


def _coerce_button_tuple(
    value: object,
    *,
    context: str,
) -> tuple[ViewerControlGamepadButtonMessage, ...]:
    if not isinstance(value, list):
        raise ViewerControlMessageError(f"{context} must be an array")

    buttons: list[ViewerControlGamepadButtonMessage] = []
    for index, button_value in enumerate(value):
        buttons.append(_coerce_gamepad_button_message(button_value, context=f"{context}[{index}]"))

    return tuple(buttons)


def _coerce_keyboard_message(value: object) -> ViewerControlKeyboardMessage:
    if not _is_plain_mapping(value):
        raise ViewerControlMessageError("keyboard must be a JSON object")

    allowed_keys = {"active_key_codes", "key_state", "focus_state", "zero_state"}
    _ensure_allowed_keys(value, allowed_keys, "keyboard")

    if "active_key_codes" not in value:
        raise ViewerControlMessageError("keyboard.active_key_codes is required")
    if "key_state" not in value:
        raise ViewerControlMessageError("keyboard.key_state is required")

    active_key_codes = _coerce_string_tuple(value["active_key_codes"], context="keyboard.active_key_codes")
    key_state = _coerce_key_state(value["key_state"], context="keyboard.key_state")
    focus_state = _coerce_optional_focus_state(value["focus_state"], context="keyboard.focus_state") if "focus_state" in value else None
    zero_state = _coerce_optional_bool(value["zero_state"], context="keyboard.zero_state") if "zero_state" in value else None

    return ViewerControlKeyboardMessage(
        active_key_codes=active_key_codes,
        key_state=key_state,
        focus_state=focus_state,
        zero_state=zero_state,
    )


def _coerce_gamepad_message(value: object) -> ViewerControlGamepadMessage:
    if not _is_plain_mapping(value):
        raise ViewerControlMessageError("gamepad must be a JSON object")

    allowed_keys = {"index", "id", "connected", "axes", "buttons", "stale", "zero_state"}
    _ensure_allowed_keys(value, allowed_keys, "gamepad")

    if "connected" not in value:
        raise ViewerControlMessageError("gamepad.connected is required")
    if "axes" not in value:
        raise ViewerControlMessageError("gamepad.axes is required")
    if "buttons" not in value:
        raise ViewerControlMessageError("gamepad.buttons is required")

    if "index" in value:
        index = value["index"]
        if not _is_int(index):
            raise ViewerControlMessageError("gamepad.index must be an integer")
        index = int(index)
    else:
        index = None

    if "id" in value:
        gamepad_id = value["id"]
        if not _is_str(gamepad_id):
            raise ViewerControlMessageError("gamepad.id must be a string")
    else:
        gamepad_id = None
    if not _is_bool(value["connected"]):
        raise ViewerControlMessageError("gamepad.connected must be a boolean")

    axes = _coerce_number_tuple(value["axes"], context="gamepad.axes")
    buttons = _coerce_button_tuple(value["buttons"], context="gamepad.buttons")
    stale = _coerce_optional_bool(value["stale"], context="gamepad.stale") if "stale" in value else None
    zero_state = _coerce_optional_bool(value["zero_state"], context="gamepad.zero_state") if "zero_state" in value else None

    return ViewerControlGamepadMessage(
        index=index,
        id=gamepad_id,
        connected=value["connected"],
        axes=axes,
        buttons=buttons,
        stale=stale,
        zero_state=zero_state,
    )


def coerce_viewer_control_message(payload: object) -> ViewerControlMessage:
    if not _is_plain_mapping(payload):
        raise ViewerControlMessageError("Invalid viewer control message: expected a JSON object")

    allowed_keys = {"type", "timestamp_s", "source_kind", "sequence", "keyboard", "gamepad", "metadata"}
    _ensure_allowed_keys(payload, allowed_keys, "viewer control message")

    if payload.get("type") != "viewer_control_message":
        raise ViewerControlMessageError("viewer control message type must be 'viewer_control_message'")
    if "timestamp_s" not in payload:
        raise ViewerControlMessageError("viewer control message.timestamp_s is required")
    if "source_kind" not in payload:
        raise ViewerControlMessageError("viewer control message.source_kind is required")

    timestamp_s = payload["timestamp_s"]
    if not _is_finite_number(timestamp_s):
        raise ViewerControlMessageError("viewer control message.timestamp_s must be a finite number")

    source_kind = payload["source_kind"]
    if source_kind not in {"keyboard", "gamepad"}:
        raise ViewerControlMessageError("viewer control message.source_kind must be 'keyboard' or 'gamepad'")

    if "sequence" in payload:
        sequence = payload["sequence"]
        if not _is_int(sequence):
            raise ViewerControlMessageError("viewer control message.sequence must be an integer")
    else:
        sequence = None

    metadata = payload["metadata"] if "metadata" in payload else {}
    if not _is_plain_mapping(metadata):
        raise ViewerControlMessageError("viewer control message.metadata must be a JSON object")
    metadata_copy: dict[str, object] = {}
    for key, value_item in metadata.items():
        if not _is_str(key):
            raise ViewerControlMessageError("viewer control message.metadata keys must be strings")
        metadata_copy[key] = value_item

    keyboard_present = "keyboard" in payload
    gamepad_present = "gamepad" in payload
    if source_kind == "keyboard":
        if not keyboard_present:
            raise ViewerControlMessageError("keyboard payload is required when source_kind is 'keyboard'")
        if gamepad_present:
            raise ViewerControlMessageError("gamepad payload is not allowed when source_kind is 'keyboard'")
        keyboard = _coerce_keyboard_message(payload["keyboard"])
        gamepad = None
    else:
        if not gamepad_present:
            raise ViewerControlMessageError("gamepad payload is required when source_kind is 'gamepad'")
        if keyboard_present:
            raise ViewerControlMessageError("keyboard payload is not allowed when source_kind is 'gamepad'")
        keyboard = None
        gamepad = _coerce_gamepad_message(payload["gamepad"])

    return ViewerControlMessage(
        type="viewer_control_message",
        timestamp_s=float(timestamp_s),
        source_kind=source_kind,
        sequence=sequence,
        keyboard=keyboard,
        gamepad=gamepad,
        metadata=metadata_copy,
    )


def parse_viewer_control_message_json(message: str) -> ViewerControlMessage:
    try:
        payload = json.loads(message)
    except json.JSONDecodeError as exc:
        raise ViewerControlMessageError("Invalid viewer control message: malformed JSON") from exc

    return coerce_viewer_control_message(payload)


__all__ = [
    "ViewerControlEnvelopeType",
    "ViewerControlGamepadButtonMessage",
    "ViewerControlGamepadMessage",
    "ViewerControlKeyboardFocusState",
    "ViewerControlKeyboardMessage",
    "ViewerControlMessage",
    "ViewerControlMessageError",
    "ViewerControlSourceKind",
    "coerce_viewer_control_message",
    "parse_viewer_control_message_json",
]

File: apps/mujoco-viewer/src/transport/viewerControlMessage.ts
export type ViewerControlSourceKind = "keyboard" | "gamepad";
export type ViewerControlEnvelopeType = "viewer_control_message";
export type ViewerControlKeyboardFocusState = "focused" | "blurred";
export type ViewerControlProviderId = "keyboard/v1" | "gamepad/v1";
export type ViewerControlProviderSchema = "viewer_keyboard_sample/v1" | "viewer_gamepad_sample/v1";

export interface ViewerControlKeyboardMessage {
  active_key_codes: string[];
  key_state: Record<string, boolean>;
  focus_state?: ViewerControlKeyboardFocusState;
  zero_state?: boolean;
}

export interface ViewerControlGamepadButtonMessage {
  pressed: boolean;
  value?: number;
}

export interface ViewerControlGamepadMessage {
  index?: number;
  id?: string;
  connected: boolean;
  axes: number[];
  buttons: ViewerControlGamepadButtonMessage[];
  stale?: boolean;
  zero_state?: boolean;
}

export interface ViewerControlMessage {
  type: ViewerControlEnvelopeType;
  timestamp_s: number;
  source_kind: ViewerControlSourceKind;
  sequence?: number;
  keyboard?: ViewerControlKeyboardMessage;
  gamepad?: ViewerControlGamepadMessage;
  metadata?: Record<string, unknown>;
  provider_id?: ViewerControlProviderId;
  provider_schema?: ViewerControlProviderSchema;
}

export class ViewerControlMessageError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ViewerControlMessageError";
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isString(value: unknown): value is string {
  return typeof value === "string";
}

function isBoolean(value: unknown): value is boolean {
  return typeof value === "boolean";
}

function isInteger(value: unknown): value is number {
  return typeof value === "number" && Number.isInteger(value);
}

function ensureAllowedKeys(value: Record<string, unknown>, allowedKeys: readonly string[], context: string): void {
  const unknownKeys = Object.keys(value).filter((key) => !allowedKeys.includes(key));
  if (unknownKeys.length > 0) {
    throw new ViewerControlMessageError(`${context} contains unknown fields: ${unknownKeys.join(", ")}`);
  }
}

function coerceStringArray(value: unknown, context: string): string[] {
  if (!Array.isArray(value) || value.some((item) => !isString(item))) {
    throw new ViewerControlMessageError(`${context} must be an array of strings`);
  }
  return value;
}

function coerceFiniteNumberArray(value: unknown, context: string): number[] {
  if (!Array.isArray(value) || value.some((item) => !isFiniteNumber(item))) {
    throw new ViewerControlMessageError(`${context} must be an array of finite numbers`);
  }
  return value;
}

function coerceKeyState(value: unknown, context: string): Record<string, boolean> {
  if (!isRecord(value)) {
    throw new ViewerControlMessageError(`${context} must be a JSON object`);
  }

  const keyState: Record<string, boolean> = {};
  for (const [key, item] of Object.entries(value)) {
    if (!isBoolean(item)) {
      throw new ViewerControlMessageError(`${context}[${JSON.stringify(key)}] must be a boolean`);
    }
    keyState[key] = item;
  }

  return keyState;
}

function coerceOptionalFocusState(value: unknown, context: string): ViewerControlKeyboardFocusState | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (value !== "focused" && value !== "blurred") {
    throw new ViewerControlMessageError(`${context} must be 'focused' or 'blurred'`);
  }
  return value;
}

function coerceOptionalBoolean(value: unknown, context: string): boolean | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (!isBoolean(value)) {
    throw new ViewerControlMessageError(`${context} must be a boolean`);
  }
  return value;
}

function coerceOptionalFiniteNumber(value: unknown, context: string): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  if (!isFiniteNumber(value)) {
    throw new ViewerControlMessageError(`${context} must be a finite number`);
  }
  return value;
}

function coerceGamepadButton(value: unknown, context: string): ViewerControlGamepadButtonMessage {
  if (!isRecord(value)) {
    throw new ViewerControlMessageError(`${context} must be a JSON object`);
  }

  ensureAllowedKeys(value, ["pressed", "value"], context);

  if (!("pressed" in value)) {
    throw new ViewerControlMessageError(`${context}.pressed is required`);
  }
  if (!isBoolean(value.pressed)) {
    throw new ViewerControlMessageError(`${context}.pressed must be a boolean`);
  }

  const button: ViewerControlGamepadButtonMessage = {
    pressed: value.pressed,
  };
  const buttonValue = coerceOptionalFiniteNumber(value.value, `${context}.value`);
  if (buttonValue !== undefined) {
    button.value = buttonValue;
  }
  return button;
}

function coerceGamepadButtons(value: unknown, context: string): ViewerControlGamepadButtonMessage[] {
  if (!Array.isArray(value)) {
    throw new ViewerControlMessageError(`${context} must be an array`);
  }

  return value.map((button, index) => coerceGamepadButton(button, `${context}[${index}]`));
}

function coerceKeyboardMessage(value: unknown): ViewerControlKeyboardMessage {
  if (!isRecord(value)) {
    throw new ViewerControlMessageError("keyboard must be a JSON object");
  }

  ensureAllowedKeys(value, ["active_key_codes", "key_state", "focus_state", "zero_state"], "keyboard");

  if (!("active_key_codes" in value)) {
    throw new ViewerControlMessageError("keyboard.active_key_codes is required");
  }
  if (!("key_state" in value)) {
    throw new ViewerControlMessageError("keyboard.key_state is required");
  }

  const keyboardMessage: ViewerControlKeyboardMessage = {
    active_key_codes: coerceStringArray(value.active_key_codes, "keyboard.active_key_codes"),
    key_state: coerceKeyState(value.key_state, "keyboard.key_state"),
  };
  const focusState = coerceOptionalFocusState(value.focus_state, "keyboard.focus_state");
  if (focusState !== undefined) {
    keyboardMessage.focus_state = focusState;
  }
  const zeroState = coerceOptionalBoolean(value.zero_state, "keyboard.zero_state");
  if (zeroState !== undefined) {
    keyboardMessage.zero_state = zeroState;
  }
  return keyboardMessage;
}

function coerceGamepadMessage(value: unknown): ViewerControlGamepadMessage {
  if (!isRecord(value)) {
    throw new ViewerControlMessageError("gamepad must be a JSON object");
  }

  ensureAllowedKeys(value, ["index", "id", "connected", "axes", "buttons", "stale", "zero_state"], "gamepad");

  if (!("connected" in value)) {
    throw new ViewerControlMessageError("gamepad.connected is required");
  }
  if (!("axes" in value)) {
    throw new ViewerControlMessageError("gamepad.axes is required");
  }
  if (!("buttons" in value)) {
    throw new ViewerControlMessageError("gamepad.buttons is required");
  }

  if (value.index !== undefined && !isInteger(value.index)) {
    throw new ViewerControlMessageError("gamepad.index must be an integer");
  }
  if (value.id !== undefined && !isString(value.id)) {
    throw new ViewerControlMessageError("gamepad.id must be a string");
  }
  if (!isBoolean(value.connected)) {
    throw new ViewerControlMessageError("gamepad.connected must be a boolean");
  }

  const gamepadMessage: ViewerControlGamepadMessage = {
    connected: value.connected,
    axes: coerceFiniteNumberArray(value.axes, "gamepad.axes"),
    buttons: coerceGamepadButtons(value.buttons, "gamepad.buttons"),
  };
  if (value.index !== undefined) {
    gamepadMessage.index = value.index;
  }
  if (value.id !== undefined) {
    gamepadMessage.id = value.id;
  }
  const stale = coerceOptionalBoolean(value.stale, "gamepad.stale");
  if (stale !== undefined) {
    gamepadMessage.stale = stale;
  }
  const zeroState = coerceOptionalBoolean(value.zero_state, "gamepad.zero_state");
  if (zeroState !== undefined) {
    gamepadMessage.zero_state = zeroState;
  }
  return gamepadMessage;
}

export function coerceViewerControlMessage(value: unknown): ViewerControlMessage {
  if (!isRecord(value)) {
    throw new ViewerControlMessageError("Invalid viewer control message: expected a JSON object");
  }

  ensureAllowedKeys(value, ["type", "timestamp_s", "source_kind", "sequence", "keyboard", "gamepad", "metadata", "provider_id", "provider_schema"], "viewer control message");

  if (value.type !== "viewer_control_message") {
    throw new ViewerControlMessageError("viewer control message type must be 'viewer_control_message'");
  }
  if (!("timestamp_s" in value)) {
    throw new ViewerControlMessageError("viewer control message.timestamp_s is required");
  }
  if (!isFiniteNumber(value.timestamp_s)) {
    throw new ViewerControlMessageError("viewer control message.timestamp_s must be a finite number");
  }
  if (!("source_kind" in value)) {
    throw new ViewerControlMessageError("viewer control message.source_kind is required");
  }
  if (value.source_kind !== "keyboard" && value.source_kind !== "gamepad") {
    throw new ViewerControlMessageError("viewer control message.source_kind must be 'keyboard' or 'gamepad'");
  }
  if (value.provider_id !== undefined && value.provider_id !== "keyboard/v1" && value.provider_id !== "gamepad/v1") {
    throw new ViewerControlMessageError("viewer control message.provider_id must be 'keyboard/v1' or 'gamepad/v1'");
  }
  if (value.provider_schema !== undefined && value.provider_schema !== "viewer_keyboard_sample/v1" && value.provider_schema !== "viewer_gamepad_sample/v1") {
    throw new ViewerControlMessageError("viewer control message.provider_schema is unknown");
  }
  const hasProviderIdentity = value.provider_id !== undefined || value.provider_schema !== undefined;
  if (hasProviderIdentity && (value.provider_id === undefined || value.provider_schema === undefined)) {
    throw new ViewerControlMessageError("viewer provider_id and provider_schema must be supplied together");
  }
  if (
    value.provider_id !== undefined &&
    value.provider_schema !== undefined &&
    ((value.source_kind === "keyboard" && (value.provider_id !== "keyboard/v1" || value.provider_schema !== "viewer_keyboard_sample/v1")) ||
      (value.source_kind === "gamepad" && (value.provider_id !== "gamepad/v1" || value.provider_schema !== "viewer_gamepad_sample/v1")))
  ) {
    throw new ViewerControlMessageError("viewer provider identity/schema does not match source_kind");
  }
  if (value.sequence !== undefined && !isInteger(value.sequence)) {
    throw new ViewerControlMessageError("viewer control message.sequence must be an integer");
  }

  const metadata = value.metadata === undefined ? undefined : value.metadata;
  if (metadata !== undefined && !isRecord(metadata)) {
    throw new ViewerControlMessageError("viewer control message.metadata must be a JSON object");
  }
  const metadataCopy = metadata === undefined ? undefined : { ...metadata };
  const sequence = value.sequence;

  if (value.source_kind === "keyboard") {
    if (value.keyboard === undefined) {
      throw new ViewerControlMessageError("keyboard payload is required when source_kind is 'keyboard'");
    }
    if (value.gamepad !== undefined) {
      throw new ViewerControlMessageError("gamepad payload is not allowed when source_kind is 'keyboard'");
    }

    const message: ViewerControlMessage = {
      type: "viewer_control_message",
      timestamp_s: value.timestamp_s,
      source_kind: "keyboard",
      keyboard: coerceKeyboardMessage(value.keyboard),
    };
    if (sequence !== undefined) {
      message.sequence = sequence;
    }
    if (metadataCopy !== undefined) {
      message.metadata = metadataCopy;
    }
    if (value.provider_id !== undefined) {
      message.provider_id = value.provider_id;
    }
    if (value.provider_schema !== undefined) {
      message.provider_schema = value.provider_schema;
    }
    return message;
  }

  if (value.gamepad === undefined) {
    throw new ViewerControlMessageError("gamepad payload is required when source_kind is 'gamepad'");
  }
  if (value.keyboard !== undefined) {
    throw new ViewerControlMessageError("keyboard payload is not allowed when source_kind is 'gamepad'");
  }

  const message: ViewerControlMessage = {
    type: "viewer_control_message",
    timestamp_s: value.timestamp_s,
    source_kind: "gamepad",
    gamepad: coerceGamepadMessage(value.gamepad),
  };
  if (sequence !== undefined) {
    message.sequence = sequence;
  }
  if (metadataCopy !== undefined) {
    message.metadata = metadataCopy;
  }
  if (value.provider_id !== undefined) {
    message.provider_id = value.provider_id;
  }
  if (value.provider_schema !== undefined) {
    message.provider_schema = value.provider_schema;
  }
  return message;
}

export function parseViewerControlMessageJson(message: string): ViewerControlMessage {
  try {
    return coerceViewerControlMessage(JSON.parse(message));
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new ViewerControlMessageError("Invalid viewer control message: malformed JSON");
    }
    throw error;
  }
}
